### ISSUE TITLE
Updates to metric_query on aws_cloudwatch_metric_alarm fail if dimensions is set

### ISSUE THREAD

**Issue by ouroboros8** [terraform-providers/terraform-provider-aws#7593]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```Terraform v0.11.11
+ provider.aws v1.59.0```

### Affected Resource(s)

<!--- Please list the affected resources and data sources. --->
* aws_cloudwatch_metric_alarm

### Terraform Configuration Files

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

```hcl
resource aws_cloudwatch_metric_alarm test {
  alarm_name                = "TEST"
  alarm_description         = "Debugging metric update in terraform"
  comparison_operator       = "GreaterThanOrEqualToThreshold"
  evaluation_periods        = "1"
  threshold                 = 10
  insufficient_data_actions = []
  ok_actions                = []
  alarm_actions             = []

  metric_query = {
    id          = "e2"
    expression  = "m1*100/m2"
    label       = "5XX percentage"
    return_data = "true"
  }

  metric_query {
    id = "m1"

    metric {
      metric_name = "5XXError"
      namespace   = "AWS/ApiGateway"
      period      = "90"
      stat        = "Sum"
      unit        = "Count"

      dimensions {
        ApiName = "Test"
        Stage   = "Test"
      }
    }
  }

  metric_query {
    id = "m2"

    metric {
      metric_name = "Count"
      namespace   = "AWS/ApiGateway"
      period      = "90"
      stat        = "Sum"
      unit        = "Count"

      dimensions {
        ApiName = "Test"
        Stage   = "Test"
      }
    }
  }
}


```

To reproduce the error, make any change to the dimension metric queries, for example change `metric_name` on `m1` to `4XXError`.

### Expected Behavior

The metric query is updates as expected to alert on percentage of 4XXs

### Actual Behavior

Apply fails with:
```Error: Error applying plan:

1 error(s) occurred:

* aws_cloudwatch_metric_alarm.test: 1 error(s) occurred:

* aws_cloudwatch_metric_alarm.test: Updating metric alarm failed: InvalidParameter: 1 validation error(s) found.
- minimum field size of 1, PutMetricAlarmInput.Metrics[3].Id.


Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

### Steps to Reproduce

1. `terraform apply` the HCL above
2. change one of the metrics, e.g. change `metric_name` on line 22 to `4XXError`
3. run `terraform apply` again

### Important Factoids

This only seems to affect metrics with `dimensions`.

**Comment by ouroboros8** [comment]
Having looked at the state, it seems that the update operation creates a 4th empty metric with no ID:
```                             "metric_query.3256485992.expression": "",
                            "metric_query.3256485992.id": "",
                            "metric_query.3256485992.label": "",
                            "metric_query.3256485992.metric.#": "0",
                            "metric_query.3256485992.return_data": "false",```
this leaves terraform in an unrecoverable state until the resource is destroyed

**Comment by sjauld** [comment]
I can confirm that the address of `PutMetricAlarmInput.Metrics[x].Id.` matches the metric query that is being blanked by terraform

**Comment by luthor2016ad** [comment]
I confirm the issue
the only way to update the alamr is to destoy it before...

**Comment by bflad** [comment]
The fix for this has been merged and will release with version 2.4.0 of the Terraform AWS Provider, likely later today.

**Comment by bflad** [comment]
This has been released in [version 2.4.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#240-march-29-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by IslamAzab** [comment]
I can replicate the error by:
- creating a relative metric
- then trying to update it

```
Updating metric alarm failed: InvalidParameter: 1 validation error(s) found.
- minimum field size of 1, PutMetricAlarmInput.Metrics[2].Id.
```

```
$ terraform -v
Terraform v0.11.13
+ provider.aws v1.60.0
```

**Comment by sjauld** [comment]
This was fixed in 2.4 as per the above. Please try upgrading!
